Step3 : 기능 우선 패키지 구성하기

MintKim1990 opened the pull request:
안녕하세요. 리뷰어님
3단계 완료하여 PR 요청드립니다 :)

2단계까지는 메뉴그룹을 하나의 바운디드컨텍스트로 나눠서 작업을 진행했었는데
메뉴그룹 자체를 놓고봤을때 비즈니스정책도 없고 특별한 역활을 가진것도 아니라 판단하여
메뉴로 합쳤습니다.

또한 패키지구조를 바운디드 컨텍스트 단위로 나눠서 작업을 하다보니
주문과 관련된 엔티티를 매장주문, 포장주문, 배달주문 3개에 엔티티로 나눴으며 각 컨텍스트마다
정책이 다르기때문에 약간에 엔티티 수정을 진행했으며 이에따라 테스트코드 또한 컨텍스트 단위로
나누고 수정진행했습니다.

이번미션에서는 패키지구조를 나누는것이 목적이다보니 테스트코드는 각 컨텍스트에 맞는 테스트만
남겨서 리팩토링 했으나 세세하게 리팩토링을 진행하진 않았습니다.

또한 비속어관련 클래스들도 메뉴, 상품 컨텍스트에 각각 구현했습니다.

중복코드가 많은거아닌가라는 의문이 약간 들었지만 바운디드컨텍스트 관점에서는 중복이라 볼수없기에
프로젝트 단위에서는 중복이 있더라도 다 찢어서 작업 진행해봤습니다

피드백 부탁드립니다 !!

rlaalsxo1990 committed:
docs: 모델링 및 용어사전 피드백사항 반영
 - 용어사전 욕설 추가
 - 상품 가격변경 모델링 설명 MenuProduct Total Price 제거
 - 설명이 모호한 부분 수정

rlaalsxo1990 committed:
docs: 모델링 및 용어사전에서 메뉴그룹 컨텍스트 위치 변경
 - 메뉴그룹 도메인이 단위 컨텍스트에서 메뉴컨텍스트로 위치 변경

rlaalsxo1990 committed:
fix: 계층 패키지구조를 바운디드 컨텍스트 단위로 패키지구조 리팩토링
 - 메뉴, 상품, 주문, 주문테이블 단위로 패키지구조 리팩토링

sah3122 commented on src/main/java/kitchenpos/deliveryorder/ui/DeliveryOrderRestController.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+import java.util.UUID;
+
+@RequestMapping("/api/orders")
```
API 정의도 order에 따라 나누는것도 좋을것 같아요 😄

sah3122 commented on src/main/java/kitchenpos/deliveryorder/domain/DeliveryOrderLineItem.java:
```diff
@@ -0,0 +1,76 @@
+import java.math.BigDecimal;
+import java.util.UUID;
+
+@Table(name = "delivery_order_line_item")
```
Entity 도 분류해주셨네요 👍 
이와 같은 상황에선 테이블을 나누는 방법도 존재하지만 Entity / Domain 을 분리하여 하나의 Entity로 사용하고 Entity <-> Domain 매핑 과정에서 유형에 따라 도메인을 분리할수도 있습니다. 
두방법의 장단점이 있으니 고민해보세요 !

MintKim1990 replied:
이부분 의도하신대로 제가 한건지 잘 모르겠네요 ㅠ

엔티티 : OrderLineItem 
도메인 : EatInOrderLineItem, TakeOutOrderLineItem, DeliveryOrderLineItem

OrderLineItem Get 하는 부분에서 도메인별로 도메인객체를 한번 매핑해서 리턴하라는 의미일까요?
혹시 한번만 더 설명해주실수 있을까요?

sah3122 replied:
이해해 주신 방법이 맞습니다 😄 
현재는 Table을 분리한 방법으로 정의해주셨는데, 위와 같은 방법도 사용할수 있고 JPA의 기능을 빌리자면 상속 매핑전략을 활용해볼수도 있을것 같아요 👍

sah3122 commented on src/main/java/kitchenpos/deliveryorder/ui/DeliveryOrderRestController.java:
```diff
@@ -0,0 +1,57 @@
+package kitchenpos.deliveryorder.ui;
```
주문 컨텍스트 내부에서 타입에 따라 패키지를 분리하는건 어떻게 생각하시나요 ?
```
- order
  - deliveryorder
  - eatinorder
  - takeoutorder
  - shared
    - vo
```
위 예시와 같이 하나의 주문 컨텍스트를 둘수 있어요. 
이와 같이 두는 이유는 일반적으로 컨텍스트는 하나의 팀만 담당하기 때문에 주문이라는 상위 컨텍스트를 두고, 하위 도메인으로 분리할수 있습니다 😄 
shared 패키지를 둔 이유는 주문에서 사용하는 공통 vo 와 같은 기능을 모아둘수 있어요. 
코멘트에서 남겨주신것 처럼 중복을 허용하지 않는다면 각 컨텍스트에서 생성할수도 있습니다 !

sah3122 commented on src/main/java/kitchenpos/menu/domain/MenuGroup.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.domain;
+package kitchenpos.menu.domain;
```
menu와 menuGroup을 함께 묶어주셨네요 👍 
패키지가 커지는것을 방지하기 위해 메뉴와 메뉴그룹을 포함하는 상위 컨텍스트를 생성하는것도 고려해보세요 😄

MintKim1990 replied:
메뉴, 메뉴그룹 도메인은 포스에 노출시킬 메뉴들과 관련된 영역으로 묶여있으니 전시라는 개념에 컨텍스트로 다시 묶어봤습니다 !

sah3122 reviewed (changes requested):
패키지 분리 잘해주셨습니다 👍 
몇가지 코멘트 남겨두었는데 확인 부탁드리며 다음 요청시 머지하도록 하겠습니다 😄

rlaalsxo1990 committed:
fix: 피드백사항 반영
 - 주문유형별 컨트롤러 Path 분기
 - 주문유형별 나눠져있던 패키지 Order 패키지로 병합
 - 메뉴 컨텍스트 전시 컨텍스르로 옮기고 메뉴, 메뉴그룹 패키지 분리

MintKim1990 commented:
리뷰어님 안녕하세요 !

피드백 주신사항 반영해보았습니다.

이해가 안가는 부분이 있어서 위에 코멘트 남겨놨는데
확인한번 부탁드립니다.

감사합니다 :)

sah3122 commented on src/main/java/kitchenpos/exhibit/menu/application/MenuService.java:
```diff
@@ -1,7 +1,11 @@
-package kitchenpos.menu.application;
+package kitchenpos.exhibit.menu.application;
```
exhibit이라는 용어가 추가됨에 따라 용어사전과 모델링문서에도 추가해주는것을 잊지 말아주세요 😄

sah3122 commented on src/main/java/kitchenpos/order/shared/domain/OrderLineItem.java:
```diff
@@ -1,14 +1,14 @@
-package kitchenpos.eatinorder.domain;
+package kitchenpos.order.shared.domain;
```
shared 활용 👍

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
질문주신 내용에 코멘트 남겨두었으니 확인 부탁드리며 이번 미션은 여기서 마무리 짓도록 할게요 🙏
